src/api/docgen/render/markdown.ts

```diff
@@ -0,0 +1,193 @@
+   */
+  readonly strike?: boolean;
+}
+
+/**
+ * Options for defining a markdown element.
+ */
+export interface MarkdownOptions {
+  /**
+   * Markdown header.
+   *
+   * @default - No header.
+   */
+  readonly header?: MarkdownHeaderOptions;
+
+  /**
+   * Id of the element.
+   *
+   * @default - The title will be used.
+   */
+  readonly id?: string;
+}
+
+/**
+ * Markdown element.
+ */
+export class Markdown {
+  /**
+   * An empty markdown element.
+   */
+  public static readonly EMPTY = new Markdown();
+
+  /**
+   * Sanitize markdown reserved characters from external input.
+   */
+  public static sanitize(line: string): string {
+    let sanitized = line.trim();
+
+    if (line.startsWith("-")) {
+      sanitized = sanitized.substring(1, line.length).trim();
+    }
+
+    return sanitized;
+  }
+
+  public static bold(text: string): string {
+    return `**${text}**`;
+  }
+
+  public static pre(text: string): string {
+    return `\`${text}\``;
+  }
+
+  public static italic(text: string) {
+    return `*${text}*`;
+  }
+
+  private readonly _lines = new Array<string>();
+  private readonly _sections = new Array<Markdown>();
+
+  private readonly id?: string;
+  private readonly header?: string;
+
+  constructor(private readonly options: MarkdownOptions = {}) {
+    this.id = options.id ?? options.header?.title;
+    this.header = this.formatHeader();
+  }
+
+  /**
+   * Render a `jsii-reflect.Docs` element into the markdown.
+   */
+  public docs(docs: reflect.Docs) {
+    if (docs.summary) {
+      this.lines(Markdown.sanitize(docs.summary));
+      this.lines("");
+    }
+    if (docs.remarks) {
+      this.lines(Markdown.sanitize(docs.remarks));
+      this.lines("");
+    }
+
+    if (docs.docs.see) {
+      this.quote(docs.docs.see);
+    }
+
+    const customLink = docs.customTag("link");
+    if (customLink) {
+      this.quote(`[${customLink}](${customLink})`);
+    }
+  }
+
+  public quote(line: string) {
+    this.lines(`> ${line}`);
+    this.lines("");
+  }
+
+  public bullet(line: string) {
+    this.lines(`- ${line}`);
+  }
+
+  public code(language: string, ...snippet: string[]) {
+    this.lines(`\`\`\`${language}`, ...snippet, "```");
+    this.lines("");
+  }
+
+  public lines(...lines: string[]) {
+    this._lines.push(...lines);
+  }
+
+  public split() {
+    this.lines("---");
+    this.lines("");
+  }
+
+  public section(section: Markdown) {
+    this._sections.push(section);
+  }
+
+  public render(headerSize: number = 0): string {
+    if (headerSize > 6) {
+      // headers are mapped to `h1-h6` html elements.
+      // passed that, markdown jsut renders `#` signs.
+      // lets see if and when we'll hit this limit.
+      throw new Error(`Unable to render markdown. Header limit (6) reached.`);
+    }
+
+    const content: string[] = [];
+    if (this.header) {
+      const heading = `${"#".repeat(headerSize)} ${this.header}`;
+      content.push(`${heading} <a name="${this.id}"></a>`);
```

Comment:
Do empty anchor tags like this have a special semantics in the markdown rendering library you're using? How would these appear to the user vs. making the entire header a link?

Reply:
They are a way to override the anchor name that is generated by default. It works both on GitHub and in the library we are using. 

> How would these appear to the user vs. making the entire header a link?

Do you mean doing something like this? 

```ts
content.push(`<a name="${this.id}">${heading} </a>`);
```

If so - this doesn't render properly:

![Screen Shot 2021-05-31 at 12 44 00 PM](https://user-images.githubusercontent.com/1428812/120174569-e78be300-c20d-11eb-9403-8df0e0e0ab4d.png)

We also don't want the entire header to be a clickable, we just need to place and anchor next to it so we can link **to** it.

Reply:
> They are a way to override the anchor name that is generated by default.

I didn't know about that, cool!